src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/MvcDetector.cs

```diff
@@ -0,0 +1,107 @@
+
+internal static class MvcDetector
+{
+    private const string ControllerTypeNameSuffix = "Controller";
+
+    // Replicates logic from ControllerFeatureProvider.IsController.
+    // https://github.com/dotnet/aspnetcore/blob/785cf9bd845a8d28dce3a079c4fedf4a4c2afe57/src/Mvc/Mvc.Core/src/Controllers/ControllerFeatureProvider.cs#L39
+    public static bool IsController(INamedTypeSymbol typeSymbol, SemanticModel semanticModel)
+    {
+        if (!typeSymbol.IsReferenceType)
+        {
+            return false;
+        }
+
+        if (typeSymbol.IsAbstract)
+        {
+            return false;
+        }
+
+        // We only consider public top-level classes as controllers.
+        if (typeSymbol.DeclaredAccessibility != Accessibility.Public)
+        {
+            return false;
+        }
+        if (typeSymbol.ContainingType != null)
+        {
+            return false;
+        }
+
+        // Has generic arguments
+        if (typeSymbol.IsGenericType)
+        {
+            return false;
+        }
+
+        // Check name before attribute's for performance.
+        if (!typeSymbol.Name.EndsWith(ControllerTypeNameSuffix, StringComparison.OrdinalIgnoreCase) &&
+            !typeSymbol.HasAttribute("Microsoft.AspNetCore.Mvc.ControllerAttribute", semanticModel))
+        {
+            return false;
+        }
+
+        if (typeSymbol.HasAttribute("Microsoft.AspNetCore.Mvc.NonControllerAttribute", semanticModel))
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    // Replicates logic from DefaultApplicationModelProvider.IsAction.
+    // https://github.com/dotnet/aspnetcore/blob/785cf9bd845a8d28dce3a079c4fedf4a4c2afe57/src/Mvc/Mvc.Core/src/ApplicationModels/DefaultApplicationModelProvider.cs#L393
+    public static bool IsAction(IMethodSymbol methodSymbol, SemanticModel semanticModel)
+    {
+        if (methodSymbol == null)
+        {
+            throw new ArgumentNullException(nameof(methodSymbol));
+        }
+
+        // The SpecialName bit is set to flag members that are treated in a special way by some compilers
+        // (such as property accessors and operator overloading methods).
+        if (methodSymbol.MethodKind is not (MethodKind.Ordinary or MethodKind.DeclareMethod))
+        {
+            return false;
+        }
+
+        // Overridden methods from Object class, e.g. Equals(Object), GetHashCode(), etc., are not valid.
+        if (methodSymbol.ContainingType.SpecialType is SpecialType.System_Object)
+        {
+            return false;
+        }
+
+        if (methodSymbol.IsStatic)
+        {
+            return false;
+        }
+
+        if (methodSymbol.IsAbstract)
+        {
+            return false;
+        }
+
+        if (methodSymbol.IsGenericMethod)
+        {
+            return false;
+        }
+
+        if (methodSymbol.DeclaredAccessibility != Accessibility.Public)
+        {
+            return false;
+        }
```

Comment:
no need to change this, but fyi you can consider doing something like:

`if (methodSymbol is { MethodKind: MethodKind.Ordinary or MethodKind.DeclareMethod, IsStatic: false, IsAbstract: false ...}) ...`

just to do this all as one simple check.  personal preference though, so up to yuou :)

Reply:
I'll leave it in its current form because it duplicates other code which uses the same syntax.